federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/MembershipType.java

```diff
@@ -96,23 +99,19 @@
                     usernames.add(username);
                 }
             } else {
-                LDAPQuery query = LDAPUtils.createQueryForUserSearch(ldapProvider, realm);
-                LDAPQueryConditionsBuilder conditionsBuilder = new LDAPQueryConditionsBuilder();
-                List<Condition> orSubconditions = new ArrayList<>();
                 for (LDAPDn userDn : dns) {
-                    String firstRdnAttrValue = userDn.getFirstRdn().getAttrValue(ldapConfig.getRdnLdapAttribute());
-                    if (firstRdnAttrValue != null) {
-                        Condition condition = conditionsBuilder.equal(ldapConfig.getRdnLdapAttribute(), firstRdnAttrValue, EscapeStrategy.DEFAULT);
-                        orSubconditions.add(condition);
+                    try {
+                        // the userDn is below the configured usersDn for sure, getLDAPMembersWithParent ensures that
+                        LDAPQuery query = LDAPUtils.createQueryForUserSearchUsingBaseDN(ldapProvider, realm, userDn.toString());
```

Comment:
@rmartinc I am not 100% sure about the performance of this as it seems that for group with 500 members, we will have 500 separate queries to LDAP if I understand correctly? By default, admin console uses 5, so performance is likely not so big concern, but admin REST api itself can be always used with much bigger value like 500 or so...

Maybe too naive approach, but won't it be sufficient to keep the original approach and just filter the duplicates as something like this?
```
    List<Integer> usernamesWithoutDuplicates = usernames.stream()
     .distinct()
     .collect(Collectors.toList());
```
This can make sure that search will return 5 users instead of 6 (In case of limit 5) and hence maybe should be fine? WDYT?

Reply:
Yep, true, it needs one search per user (obtaining groups of a user is one search but users of a group is N searches, ldap works this way). We cannot filter usernames because the problem is with the CN, but, in order to the same that is done now, we can check that the returned DN is in the initial set (not another user with the same CN but not member of the group). Something like the following:

```java
                for (LDAPObject ldapUser : ldapUsers) {
                    if (dns.contains(ldapUser.getDn()) {
                        String username = LDAPUtils.getUsername(ldapUser, ldapConfig);
                        usernames.add(username);
                    }
                }
```

But I think that creating an OR filter with 500 operands is not very good either. But we are already doing that way so we can just maintain it if you prefer. I can use a lambda too for this.

Reply:
@mposolda do I prepare that change?

Reply:
+1 to stick with OR and do like the `for` cycle or lambda as you proposed